File: rust/fory-core/src/serializer/serializer.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::error::Error;
use crate::meta::FieldInfo;
use crate::resolver::context::{ReadContext, WriteContext};
use crate::serializer::util::{
    read_compatible_default, read_ext_type_info, read_ref_info_data, write_ext_type_info,
    write_ref_info_data,
};
use crate::serializer::{bool, struct_};
use crate::types::TypeId;
use crate::TypeResolver;
use std::any::Any;

pub trait ForyDefault: Sized {
    fn fory_default() -> Self;
}

// We can't add blanket impl for all T: Default because it conflicts with other impls.
// For example, upstream crates may add a new impl of trait `std::default::Default` for
// type `std::rc::Rc<(dyn std::any::Any + 'static)>` in future versions.
// impl<T: Default + Sized> ForyDefault for T {
//     fn fory_default() -> Self {
//         Default::default()
//     }
// }

pub trait Serializer: 'static {
    /// Entry point of the serialization.
    fn fory_write(&self, context: &mut WriteContext) -> Result<(), Error>
    where
        Self: Sized,
    {
        write_ref_info_data(self, context, false, false)
    }

    fn fory_read(context: &mut ReadContext) -> Result<Self, Error>
    where
        Self: Sized + ForyDefault,
    {
        read_ref_info_data(context, false, false)
    }

    fn fory_is_option() -> bool
    where
        Self: Sized,
    {
        false
    }

    fn fory_is_none(&self) -> bool {
        false
    }

    fn fory_is_polymorphic() -> bool
    where
        Self: Sized,
    {
        false
    }

    fn fory_is_shared_ref() -> bool
    where
        Self: Sized,
    {
        false
    }

    fn fory_static_type_id() -> TypeId
    where
        Self: Sized,
    {
        // set to ext to simplify the user defined serializer.
        // serializer for other types will override this method.
        TypeId::EXT
    }

    fn fory_get_type_id(type_resolver: &TypeResolver) -> Result<u32, Error>
    where
        Self: Sized,
    {
        Ok(type_resolver
            .get_type_info(std::any::TypeId::of::<Self>())?
            .get_type_id())
    }

    fn fory_type_id_dyn(&self, type_resolver: &TypeResolver) -> Result<u32, Error>;

    /// The possible max memory size of the type.
    /// Used to reserve the buffer space to avoid reallocation, which may hurt performance.
    fn fory_reserved_space() -> usize
    where
        Self: Sized,
    {
        0
    }

    fn fory_write_type_info(context: &mut WriteContext) -> Result<(), Error>
    where
        Self: Sized,
    {
        // default implementation only for ext/named_ext
        // this method will be overridden by serializers for supported types
        write_ext_type_info::<Self>(context)
    }

    fn fory_read_type_info(context: &mut ReadContext) -> Result<(), Error>
    where
        Self: Sized,
    {
        // default implementation only for ext/named_ext
        // this method will be overridden by serializers for supported types
        read_ext_type_info::<Self>(context)
    }

    // only used by struct/enum/ext
    fn fory_read_compatible(context: &mut ReadContext) -> Result<Self, Error>
    where
        Self: Sized,
    {
        // default logic only for ext/named_ext
        // this method will be overridden by fory macro generated serializers
        read_compatible_default::<Self>(context)
    }

    /// Write the data into the buffer. Need to be implemented.
    fn fory_write_data(&self, context: &mut WriteContext) -> Result<(), Error>;

    /// Write the data into the buffer. Need to be implemented for collection/map.
    /// For other types, just forward to `fory_write_data`.
    fn fory_write_data_generic(
        &self,
        context: &mut WriteContext,
        _has_generics: bool,
    ) -> Result<(), Error> {
        self.fory_write_data(context)
    }

    fn fory_read_data(context: &mut ReadContext) -> Result<Self, Error>
    where
        Self: Sized + ForyDefault;

    fn fory_concrete_type_id(&self) -> std::any::TypeId {
        std::any::TypeId::of::<Self>()
    }

    fn as_any(&self) -> &dyn Any;
}

pub trait StructSerializer: Serializer + 'static {
    fn fory_fields_info(_: &TypeResolver) -> Result<Vec<FieldInfo>, Error> {
        Ok(Vec::default())
    }

    fn fory_type_index() -> u32 {
        unimplemented!()
    }

    fn fory_actual_type_id(type_id: u32, register_by_name: bool, compatible: bool) -> u32 {
        struct_::actual_type_id(type_id, register_by_name, compatible)
    }

    fn fory_get_sorted_field_names() -> &'static [&'static str] {
        &[]
    }
}

pub trait CollectionSerializer: Serializer + 'static {
    fn fory_write_collection_field(&self, context: &mut WriteContext) -> Result<(), Error>;
}

pub trait MapSerializer: Serializer + 'static {
    fn fory_write_map_field(&self, context: &mut WriteContext) -> Result<(), Error>;
}

File: swift/Sources/Fory/Serializer.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

public protocol Serializer {
    static func foryDefault() -> Self
    static var staticTypeId: TypeId { get }

    static var isNullableType: Bool { get }
    static var isRefType: Bool { get }

    var foryIsNone: Bool { get }

    func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws
    static func foryReadData(_ context: ReadContext) throws -> Self

    func foryWrite(
        _ context: WriteContext,
        refMode: RefMode,
        writeTypeInfo: Bool,
        hasGenerics: Bool
    ) throws

    static func foryRead(
        _ context: ReadContext,
        refMode: RefMode,
        readTypeInfo: Bool
    ) throws -> Self

    static func foryWriteStaticTypeInfo(_ context: WriteContext) throws
    static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo?
    static func foryReadCompatibleData(_ context: ReadContext, remoteTypeInfo: TypeInfo) throws -> Self
    static func foryFieldsInfo(trackRef: Bool) -> [TypeMeta.FieldInfo]
    func foryWriteTypeInfo(_ context: WriteContext) throws
}

public protocol StructSerializer: Serializer {
    static var foryEvolving: Bool { get }
}

public extension Serializer {
    @inlinable
    static var isNullableType: Bool { false }

    @inlinable
    static var isRefType: Bool { false }

    @inlinable
    var foryIsNone: Bool { false }

    @inlinable
    func foryWriteTypeInfo(_ context: WriteContext) throws {
        try Self.foryWriteStaticTypeInfo(context)
    }

    @inlinable
    static func foryFieldsInfo(trackRef _: Bool) -> [TypeMeta.FieldInfo] {
        []
    }

    @inlinable
    static func foryReadCompatibleData(_ context: ReadContext, remoteTypeInfo _: TypeInfo) throws -> Self {
        try foryReadData(context)
    }
}

public extension StructSerializer {
    @inlinable
    static var foryEvolving: Bool { true }
}

public extension Serializer {
    static func foryReadPayload(
        _ context: ReadContext,
        readTypeInfo: Bool
    ) throws -> Self {
        try Self.foryReadPayload(
            context,
            readTypeInfo: readTypeInfo,
            readData: { try Self.foryReadData(context) },
            readCompatibleData: { remoteTypeInfo in
                try Self.foryReadCompatibleData(context, remoteTypeInfo: remoteTypeInfo)
            }
        )
    }

    static func foryReadPayload(
        _ context: ReadContext,
        readTypeInfo: Bool,
        readData: () throws -> Self,
        readCompatibleData: (TypeInfo) throws -> Self
    ) throws -> Self {
        if readTypeInfo {
            if let remoteTypeInfo = try Self.foryReadTypeInfo(context) {
                return try readCompatibleData(remoteTypeInfo)
            }
        } else if let remoteTypeInfo = context.getTypeInfo(for: Self.self) {
            return try readCompatibleData(remoteTypeInfo)
        }
        return try readData()
    }

    @inlinable
    func foryWrite(
        _ context: WriteContext,
        refMode: RefMode,
        writeTypeInfo: Bool,
        hasGenerics: Bool
    ) throws {
        if refMode != .none {
            if refMode == .tracking, Self.isRefType, let object = self as AnyObject? {
                if context.refWriter.tryWriteRef(buffer: context.buffer, object: object) {
                    return
                }
            } else {
                context.buffer.writeInt8(RefFlag.notNullValue.rawValue)
            }
        }

        if writeTypeInfo {
            try Self.foryWriteStaticTypeInfo(context)
        }

        try foryWriteData(context, hasGenerics: hasGenerics)
    }

    @inlinable
    static func foryRead(
        _ context: ReadContext,
        refMode: RefMode,
        readTypeInfo: Bool
    ) throws -> Self {
        switch refMode {
        case .none:
            return try Self.foryReadPayload(context, readTypeInfo: readTypeInfo)
        case .nullOnly:
            let rawFlag = try context.buffer.readInt8()
            switch rawFlag {
            case RefFlag.null.rawValue:
                return Self.foryDefault()
            case RefFlag.notNullValue.rawValue:
                return try Self.foryReadPayload(context, readTypeInfo: readTypeInfo)
            case RefFlag.refValue.rawValue:
                if context.trackRef {
                    let reservedRefID = context.refReader.reserveRefID()
                    let value = try Self.foryReadPayload(context, readTypeInfo: readTypeInfo)
                    if let object = value as AnyObject? {
                        context.refReader.storeRef(object, at: reservedRefID)
                    }
                    return value
                }
                return try Self.foryReadPayload(context, readTypeInfo: readTypeInfo)
            case RefFlag.ref.rawValue:
                let refID = try context.buffer.readVarUInt32()
                return try context.refReader.readRef(refID, as: Self.self)
            default:
                throw ForyError.refError("invalid ref flag \(rawFlag)")
            }
        case .tracking:
            let rawFlag = try context.buffer.readInt8()
            guard let flag = RefFlag(rawValue: rawFlag) else {
                throw ForyError.refError("invalid ref flag \(rawFlag)")
            }
            switch flag {
            case .null:
                return Self.foryDefault()
            case .ref:
                let refID = try context.buffer.readVarUInt32()
                return try context.refReader.readRef(refID, as: Self.self)
            case .refValue:
                let reservedRefID = context.trackRef ? context.refReader.reserveRefID() : nil
                let value = try Self.foryReadPayload(context, readTypeInfo: readTypeInfo)
                if let reservedRefID, let object = value as AnyObject? {
                    context.refReader.storeRef(object, at: reservedRefID)
                }
                return value
            case .notNullValue:
                return try Self.foryReadPayload(context, readTypeInfo: readTypeInfo)
            }
        }
    }

    static func foryWriteStaticTypeInfo(_ context: WriteContext) throws {
        guard staticTypeId.isUserTypeKind else {
            context.buffer.writeUInt8(UInt8(truncatingIfNeeded: staticTypeId.rawValue))
            return
        }

        let typeInfo = try context.typeInfo(for: Self.self)
        let wireTypeID = typeInfo.wireTypeID(compatible: context.compatible)
        context.buffer.writeUInt8(UInt8(truncatingIfNeeded: wireTypeID.rawValue))
        switch wireTypeID {
        case .compatibleStruct, .namedCompatibleStruct:
            guard typeInfo.typeDefBytes != nil else {
                throw ForyError.invalidData("missing compatible type definition for \(typeInfo.typeID)")
            }
            context.writeTypeMeta(typeInfo)
        case .namedEnum, .namedStruct, .namedExt, .namedUnion:
            if context.compatible {
                guard typeInfo.typeDefBytes != nil else {
                    throw ForyError.invalidData("missing compatible type definition for \(typeInfo.typeID)")
                }
                context.writeTypeMeta(typeInfo)
            } else {
                try writeMetaString(
                    context: context,
                    value: typeInfo.namespace,
                    encodings: namespaceMetaStringEncodings,
                    encoder: .namespace
                )
                try writeMetaString(
                    context: context,
                    value: typeInfo.typeName,
                    encodings: typeNameMetaStringEncodings,
                    encoder: .typeName
                )
            }
        default:
            if !typeInfo.registerByName && registeredWireTypeNeedsUserTypeID(wireTypeID) {
                guard let userTypeID = typeInfo.userTypeID else {
                    throw ForyError.invalidData("missing user type id for id-registered type")
                }
                context.buffer.writeVarUInt32(userTypeID)
            }
        }
    }

    static func foryReadTypeInfo(_ context: ReadContext) throws -> TypeInfo? {
        try context.readTypeInfo(for: Self.self)
    }

}
